File: src/models/Campaign.js
const mongoose = require('mongoose');

const campaignSchema = new mongoose.Schema({
    title: {
        type: String,
        required: [true, 'Campaign title is required'],
        trim: true
    },
    description: {
        type: String,
        required: [true, 'Campaign description is required']
    },
    creator: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    target_amount: {
        type: Number,
        required: [true, 'Target amount is required'],
        min: [1, 'Target amount must be greater than 0']
    },
    current_amount: {
        type: Number,
        default: 0
    },
    status: {
        type: String,
        enum: ['active', 'completed', 'cancelled'],
        default: 'active'
    },
    deadline: {
        type: Date,
        required: [true, 'Campaign deadline is required']
    },
    category: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Category',
        required: true
    },
    images: [{
        type: String
    }],
    donors: [{
        user: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'User'
        },
        amount: Number,
        date: {
            type: Date,
            default: Date.now
        }
    }],
    comments: [{
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Comment'
    }]
}, {
    timestamps: true,
    toJSON: { virtuals: true },
    toObject: { virtuals: true }
});

// Virtual field for progress percentage
campaignSchema.virtual('progress').get(function() {
    return (this.current_amount / this.target_amount) * 100;
});

// Virtual field for days left
campaignSchema.virtual('days_left').get(function() {
    const now = new Date();
    const timeDiff = this.deadline.getTime() - now.getTime();
    return Math.ceil(timeDiff / (1000 * 3600 * 24));
});

// Update campaign status based on amount and deadline
campaignSchema.methods.updateStatus = function() {
    const now = new Date();
    if (this.current_amount >= this.target_amount) {
        this.status = 'completed';
    } else if (now > this.deadline) {
        this.status = 'cancelled';
    }
};

module.exports = mongoose.model('Campaign', campaignSchema);

File: models/campaign.py
from datetime import datetime
from bson import ObjectId

class Campaign:
    def __init__(self, title, description, goal, creator_id):
        self.title = title
        self.description = description
        self.goal = float(goal)
        self.creator_id = creator_id
        self.current_amount = 0.0
        self.status = 'active'
        self.created_at = datetime.utcnow()
        self.updated_at = datetime.utcnow()
        self.donors = []
        self.images = []
        self.updates = []
        self.comments = []
    
    def to_dict(self):
        return {
            'title': self.title,
            'description': self.description,
            'goal': self.goal,
            'creator_id': self.creator_id,
            'current_amount': self.current_amount,
            'status': self.status,
            'created_at': self.created_at,
            'updated_at': self.updated_at,
            'donors': self.donors,
            'images': self.images,
            'updates': self.updates,
            'comments': self.comments
        }
    
    @staticmethod
    def from_dict(data):
        campaign = Campaign(
            title=data['title'],
            description=data['description'],
            goal=data['goal'],
            creator_id=data['creator_id']
        )
        campaign.current_amount = data.get('current_amount', 0.0)
        campaign.status = data.get('status', 'active')
        campaign.created_at = data.get('created_at', datetime.utcnow())
        campaign.updated_at = data.get('updated_at', datetime.utcnow())
        campaign.donors = data.get('donors', [])
        campaign.images = data.get('images', [])
        campaign.updates = data.get('updates', [])
        campaign.comments = data.get('comments', [])
        return campaign
